Feature: Timer and gameover  #19

sourabhaa opened the pull request:
A timer of 1 minute is set for each instance of game. After te end of each game, the total score and  high score is displayed. Comparing the score and high score, win/lose of each game is decided.

![Screenshot of  JAMath Activity](https://user-images.githubusercontent.com/43743186/113333391-00783500-9340-11eb-96d7-a0893d14e8dd.png)

![Screenshot of  JAMath Activity](https://user-images.githubusercontent.com/43743186/113333553-2ef61000-9340-11eb-84b8-2dbda0a91508.png)


- Majority of the code and the text displayed is in Spanish. Since the activity is available to all the Sugar users, I'd like to change the texts to English. 
-  There is a glitch in storing and fetching the high score from the journal, hence the high score is always shown as 0. I'll try to figure it out soon.
- Colour of the background image is too bright due to which it is hard to read the content.   
 
Thoughts and suggestions re this are welcome. Also, should I raise issues for these or fix it in this PR?

chimosky commented:
Haven't reviewed or tested yet but could you do a rebase and drop dcd8365 and f9d8630?

sourabhaa commented:
Sorry..force push caused the PR to close!  @chimosky could you please reopen it?

sourabhaa commented:
I rebased and pushed the changes. Commits include,
1. Added timer
2. gameover page
3. fixed high score glitch
4. added ```play-again``` and ```exit``` button in game over page.
5. fixed minor pep8 issues
 
UI of ```game over``` doesn't look so attractive now because of the background image. I could fade the image a little or change. I'll proceed as per your suggestions. @chimosky kindly review.

chimosky reviewed (commented):
What branch is 87b6f24 a rebase of?

Reviewed, not tested.

chimosky commented on juego.py:
```diff
@@ -394,18 +395,22 @@ def run(self):
         sonido_menu = load_sound("menu.ogg")
         gameover = self.fuente_130.render("GAME OVER!!", True, (255, 0, 0))
         quit = self.fuente_130.render("EXIT", True, (0, 0, 0))
+        play_again = self.fuente_130.render("play again", True, (0, 0, 0))
```
Would be nice if "play again" was all caps.

sourabhaa replied:
> Would be nice if "play again" was all caps.

Sure, I missed it (but the font isn't case sensitive, everything reflects to be in caps when rendered)

> What branch is [87b6f24](https://github.com/sugarlabs/jamath-activity/commit/87b6f24772547ef344ae0bd9214394ad1c48a2ba) a rebase of?
> 
> Reviewed, not tested.

master (before the merge of #21). In next commit, I'll pull these changes.

chimosky replied:
> Sure, I missed it (but the font isn't case sensitive, everything reflects to be in caps when rendered)

Oh okay.

chimosky commented on juego.py:
```diff
@@ -484,9 +496,11 @@ def game_over(self, score, high_score):
                                 quit.get_width() and \
-                                event.pos[1] > sy(700) and event.pos[1] < sy(700) + \
+                                event.pos[1] > sy(700) and \
+                                event.pos[1] < sy(700) + \
```
For readability and pep8 standard, the operators should be on the left of the operands.

sourabhaa replied:
Sure, will fix it.

chimosky commented:
Tested 174cb4b, activity fails to start with error;
```
Traceback (most recent call last):
  File "/usr/bin/sugar-activity3", line 5, in <module>
    activityinstance.main()
  File "/usr/lib/python3.7/dist-packages/sugar3/activity/activityinstance.py", line 179, in main
    module = __import__(module_name)
  File "/home/ibiam/Activities/jamath-activity/JAMath.py", line 10, in <module>
    from juego import Game
  File "/home/ibiam/Activities/jamath-activity/juego.py", line 523
    if event.pos[0] > sx(800) and \
                                  ^
IndentationError: unindent does not match any outer indentation level
```

Also, can you try avoiding merge commits, you can do a rebase instead.

sourabhaa commented:
> Tested [174cb4b](https://github.com/sugarlabs/jamath-activity/commit/174cb4bd8926134d208177db0c997636cf0b29fe), activity fails to start with error;

Sorry, fixed it.

chimosky commented on juego.py:
```diff
@@ -386,11 +399,138 @@ def run(self):
+            score, puntuacionalta = self.play(level)
+            play_again = self.game_over(score, puntuacionalta)
+
+    def game_over(self, score, high_score):
```
There seems to be a repetition of what's been done earlier by `play` like checking for events and handling those events and `MOUSEBUTTONDOWN` seems to be handled by both and `game_over` contains lots of code that I feel it doesn't really need. Why not just render the necessary text when the game is over?

sourabhaa replied:
I removed unnecessary code. But I'd need to use `MOUSEBUTTONDOWN` as there are `Play-again` and `Exit` buttons in `game_over`. Also, I deleted this branch by mistake. So will it be okay to make another PR with changes and close this one? Or is there a way to push commits from a new branch to this PR?

quozl commented:
While the pull request remains, you or anyone can "git fetch sugarlabs
pull/N/head" and "git checkout M" where N is the pull request number,
and M is the commit hash you want to look at.

Should you have used a different remote name for the sugarlabs
repository, change sugarlabs in the fetch.

sourabhaa commented:
>While the pull request remains, you or anyone can "git fetch sugarlabs
pull/N/head" and "git checkout M" where N is the pull request number,
and M is the commit hash you want to look at.

Thanks @quozl , but that is not the case here, I messed up a little with pulling and rebasing the code from the remote branch and ended up deleting the repo by mistake, hence I won't be able to access this PR. I will create a new PR and reference this PR in the new one.

quozl commented:
That's really odd.  I'm able to access this pull request.  I am able
to checkout 30030ecf1b520422fae5592d446fda43e3a8acbb.  Deleting my
clone won't change that, as I can just fetch and checkout again.
Either you didn't explain what you needed, or you are mistaken.

chimosky commented:
@sourabhaa said;
> Thanks @quozl , but that is not the case here, I messed up a little with pulling and rebasing the code from the remote branch and ended up deleting the repo by mistake, hence I won't be able to access this PR. I will create a new PR and reference this PR in the new one.

Could you explain what you did? It'll be nice if you can tell everything, from pulling to rebasing and deleting.

sourabhaa commented:
Since I tried a lot of ways, I'm not sure about what exactly I messed up and if this explanation is apt. But This is what I think I did. I had made few changes to the local branch and then rebased the updated code from the remote branch. But it wouldn't merge the changes. It would ask to manually merge each change and it would show those merged changes as fresh change in the current commit. So I deleted the local branch and pulled everything from remote branch...but then, the master would also have adapted those changes. I hadn't realised this and made commits to master and instead of pushing the changes to branch, I used UI of VS code which pushed and pulled all changes in all branches. The changes in the repo was confusing, the master branch was a lot of commits ahead and behind the master of Sugarlabs and the changes in each commits were repeated as fresh commits in each commit. So I deleted the repo and started afresh. After deleting the repo, I cannot fetch the code from this PR or push the changes from other branch to this PR. It takes as a fresh PR.

chimosky commented:
If such happens next time, you can do this;
```bash
# This resets your local master to the current master at HEAD, you can also use git switch. 
$ git reset --hard bd72848a  
```

So you have your local master now at the remote master and you can fetch a branch you don't intend
to merge into the correct branch and checkout that branch;
```
$ git fetch remote branch
$ git checkout branch
```

Also, after deleting this branch locally you can also checkout out to the deleted branch again,
as long as you'd fetched it at some point from your remote and it'll checkout to your new branch.

When you started afresh, I'm guessing you cloned this repo so there's a probability you forgot to 
add your fork as a remote and pull from that as that's the only reason why you wouldn't be able to
do so - speculating here as I didn't see any output you got from git when trying to fetch changes from
your branch -.

There's a tool called `script` that you can use when you're trying out something new and there's a 
probability you won't remember all the commands you typed, comes in handy.

quozl commented:
Couple of other tools for looking at what happened to lead you to a
place where there are dragons;

1.  "git reflog",

2.  bash history.

Daily or hourly backups can also reduce doubt.

chimosky commented:
Closing this as it was reopened in #23.
